ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/CSJ.hs

```diff
@@ -0,0 +1,134 @@
+                     condenseListWithPadding)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Test.Consensus.BlockTree (BlockTree (..))
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.Genesis.Tests.Uniform (genUniformSchedulePoints)
+import           Test.Consensus.PeerSimulator.Run (SchedulerConfig (..),
+                     defaultSchedulerConfig)
+import           Test.Consensus.PeerSimulator.StateView (StateView (..))
+import           Test.Consensus.PeerSimulator.Trace (TraceEvent (..))
+import           Test.Consensus.PointSchedule
+import           Test.Consensus.PointSchedule.Peers (Peer (..), Peers (..),
+                     mkPeers)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (Header, TestBlock)
+import           Test.Util.TestEnv (adjustQuickCheckMaxSize)
+
+tests :: TestTree
+tests =
+  adjustQuickCheckMaxSize (`div` 5) $
+  testGroup
+    "CSJ"
+    [ testGroup "Happy Path"
+      [ testProperty "synchronous" $ prop_happyPath True
+      , testProperty "asynchronous" $ prop_happyPath False
+      ]
+    ]
+
+-- | Test of the “happy path” scenario of ChainSync Jumping (CSJ).
+--
+-- This test features one chain (ie. a block tree that is only trunk) and only
+-- honest peers and syncs the chain in question with CSJ enabled. What we expect
+-- to observe is that one of the honest peers becomes the dynamo while the
+-- others become jumpers. Because the jumpers will agree to all the jumps, the
+-- whole syncing should happen with CSJ without objectors.
+--
+-- The final property is that headers should only ever be downloaded once and
+-- only from one peer (the dynamo). This is true except when almost caught-up:
+-- when the dynamo is caught-up, it gets disengaged and one of the jumpers takes
+-- its place and starts serving headers. This might lead to duplication of
+-- headers, but only in a window of @jumpSize@ slots near the tip of the chain.
+--
+-- The boolean differentiates between “synchronous” and “asynchronous”
+-- scenarios. In a synchronous scenario, all the honest peers have the same
+-- schedule: they serve the chain exactly in the same way. In the asynchronous
+-- scenario, a random schedule is generated for each peer (but they still serve
+-- the same chain).
+prop_happyPath :: Bool -> Property
+prop_happyPath synchronized =
+  forAllGenesisTest
+    ( do
+        gt <- genChains $ pure 0
+        honest <- genHonestSchedule gt
+        numOthers <- choose (1, 3)
+        otherHonests <- if synchronized
+          then pure $ replicate numOthers honest
+          else replicateM numOthers (genHonestSchedule gt)
+        pure $ gt $> mkPeers honest otherHonests
+    )
+    ( defaultSchedulerConfig
+      { scEnableCSJ = True
+      , scEnableLoE = True
+      , scEnableLoP = True
```

Comment:
It is possible to test with CSJ without LoE or LoP. I'm guessing this scenario would not be too interesting as CSJ is only going to be used during syncing.

Reply:
It's fine as is, I think.

Reply:
The “happy path” without adversaries can be tested without LoE or LoP, indeed. The version with adversaries probably shouldn't because it is possible that an adversary, as dynamo or objector, serves headers fast enough for the honest peers (dynamo, objector, or jumpers) to fall behind. I say “probably” because the property of that test does not actually enforce that we end up on the honest chain, so maybe the test would still technically pass.

We could slightly generalise the test by running the happy path with and without LoE+LoP, and the non-happy path always with LoE+LoP?